topics/go/language/embedding/exercises/exercise1/exercise1.go

```diff
@@ -1,64 +1,110 @@
-// for the embedded sport type to check the embedded fields first. Then create
-// two hockey values inside the slice of finders and perform the search.
+// Sample program to show how you can use embedding to reuse behavior from
+// another type and override specific methods.
 package main
 
 import (
 	"fmt"
-	"strings"
+	"log"
+	"time"
 )
 
-// finder defines the behavior required for performing matching.
-type finder interface {
-	find(needle string) bool
+// Document is the core data model we are working with.
+type Document struct {
+	Key   string
+	Title string
 }
 
-// sport represents a sports team.
-type sport struct {
-	team string
-	city string
-}
+// ==================================================
+
+// Feed is a type that knows how to fetch Documents.
+type Feed struct{}
 
-// find checks the value for the specified term.
-func (s *sport) find(needle string) bool {
-	return strings.Contains(s.team, needle) || strings.Contains(s.city, needle)
+// Count tells how many documents are in the feed.
+func (f *Feed) Count() int {
+	return 42
 }
 
-// hockey represents specific hockey information.
-type hockey struct {
-	sport
-	country string
+// Fetch simulates looking up the document specified by key. It is slow.
+func (f *Feed) Fetch(key string) (Document, error) {
+
+	time.Sleep(1000 * time.Millisecond)
+
+	doc := Document{
+		Key:   key,
+		Title: "Title for " + key,
+	}
+	return doc, nil
 }
 
-// find checks the value for the specified term.
-func (h *hockey) find(needle string) bool {
-	return h.sport.find(needle) || strings.Contains(h.country, needle)
+// ==================================================
+
+// CachingFeed keeps a local copy of Documents that have already been
+// retrieved. It embeds Feed to get the Fetch and Count behavior but
+// "overrides" Fetch to have its cache.
+type CachingFeed struct {
+	docs map[string]Document
+	*Feed
 }
 
-func main() {
+// NewCachingFeed initalizes a CachingFeed for use.
+func NewCachingFeed(f *Feed) *CachingFeed {
+	fc := CachingFeed{
+		docs: make(map[string]Document),
+		Feed: f,
+	}
+	return &fc
+}
 
-	// Define the term to find.
-	needle := "Miami"
+// Fetch calls the embedded type's Fetch method if the key is not cached.
+func (r *CachingFeed) Fetch(key string) (Document, error) {
+	doc, ok := r.docs[key]
+	if ok {
+		return doc, nil
+	}
 
-	// Create a slice of finder values and assign values
-	// of the concrete hockey type.
-	haystack := []finder{
-		&hockey{sport{"Panthers", "Miami"}, "USA"},
-		&hockey{sport{"Canadiens", "Montreal"}, "Canada"},
+	doc, err := r.Feed.Fetch(key)
+	if err != nil {
+		return Document{}, err
 	}
 
-	// Display what we are trying to find.
-	fmt.Println("Searching For:", needle)
+	r.docs[key] = doc
+	return doc, nil
+}
+
+// ==================================================
+
+// FetchCounter is the behavior we depend on for our process function.
+type FetchCounter interface {
```

Comment:
Don't understand the need for this interface? There is only one implementation. is this necessary?

Reply:
There will be two implementations `Feed` and `CachingFeed`. In the final solution we will pass one of each of these into `process`.